ray-curator/ray_curator/stages/deduplication/semantic/pairwise.py

```diff
@@ -0,0 +1,326 @@
+            ranking_strategy: Strategy for ranking/sorting clusters before similarity computation.
+            pairwise_batch_size: Batch size for pairwise similarity computation.
+            verbose: Whether to print verbose output.
+            embedding_dim: Embedding dimension for memory estimation.
+            read_kwargs: Kwargs for reading parquet files.
+            write_kwargs: Kwargs for writing parquet files.
+        """
+        self.id_field = id_field
+        self.embedding_field = embedding_field
+        self.output_path = output_path
+        self.pairwise_batch_size = pairwise_batch_size
+        self.embedding_dim = embedding_dim
+        self.ranking_strategy = ranking_strategy
+        self.verbose = verbose
+        self.read_kwargs = read_kwargs if read_kwargs is not None else {}
+        self.write_kwargs = write_kwargs if write_kwargs is not None else {}
+        self.input_storage_options = self.read_kwargs.pop("storage_options", None) if self.read_kwargs else None
+        self.output_storage_options = self.write_kwargs.pop("storage_options", None) if self.write_kwargs else None
+        self._name = "PairwiseCosineSimilarityStage"
+        self._resources = Resources(cpus=1.0, gpus=1.0)
+
+    def process(self, task: FileGroupTask) -> FileGroupTask:
+        """Process a PairwiseFileGroupTask to compute pairwise similarities."""
+        if task._metadata.get("filetype") != "parquet":
+            msg = f"PairwiseCosineSimilarityStage only supports parquet files, got {task._metadata.get('filetype')}"
+            raise ValueError(msg)
+
+        cluster_id = task._metadata.get("centroid_id")
+        output_path = os.path.join(self.output_path, f"cluster_{cluster_id}.parquet")
+        if cluster_id is None:
+            msg = "centroid_id not found in task metadata"
+            raise ValueError(msg)
+
+        t1 = time.perf_counter()
+
+        # Read all file groups and concatenate
+        dfs = []
+        num_rows = 0
+
+        # Break input files into groups to avoid 2bn row limit
+        file_groups = break_parquet_partition_into_groups(
+            task.data, embedding_dim=self.embedding_dim, storage_options=self.input_storage_options
+        )
+
+        # Determine which columns to read based on ranking strategy
+        additional_cols = self.ranking_strategy.metadata_cols if self.ranking_strategy.strategy == "sort" else []
+
+        # We do the list(dict.fromkeys(...)) to remove duplicates from the list of columns to read, in case additional_cols contains self.id_field
+        metadata_cols = list(dict.fromkeys([self.id_field, *additional_cols]))
+        for file_group in file_groups:
+            # Read required columns including metadata columns for ranking
+            df = self.read_parquet(
+                file_group,
+                columns=[*metadata_cols, self.embedding_field],
+                assign_id=False,
+                storage_options=self.input_storage_options,
```

Comment:
So we may be providing storage options both as part of FileGroupTask and read_args. For now i am giving precedence of FileGroupTask . Wonder if we should figure out a way to force this

Reply:
I don't think FileGroupTask should have storage_options